test/cluster/basic_test.go

```diff
@@ -101,3 +101,348 @@ func TestBasicCluster(t *testing.T) {
+}
+
+// Scenario:
+// - start 3 servers in a cluster.
+// - shutting down and starting all servers in the cluster.
+func TestClusterRecovery(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	require.NoError(t, c.Restart())
+}
+
+// Scenario:
+// - start 3 servers in a cluster.
+// - wait for one to be the leader.
+// - submit a tx to each of the 3, with a client that catches redirects.
+// - stop the leader.
+// - wait for new server to be the leader.
+// - repeat until each server was a leader at least once.
+func TestAllNodesGetLeadership(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	time.Sleep(10 * time.Second) //TODO replace by a call to detect a leader
+
+	var txOK int
+	leaders := [3]int{0, 0, 0}
+	currentLeader := -1
+	oldLeader := -1
+
+	for leaders[0] == 0 || leaders[1] == 0 || leaders[2] == 0 {
+		time.Sleep(10 * time.Second)
+		for i, srv := range c.Servers {
+			if i != currentLeader {
+				txID, rcpt, err := srv.WriteDataTx(t, worldstate.DefaultDBName, "alice", []byte{uint8(i), uint8(i)})
+				if err != nil {
+					t.Logf("txID: %s, error: %s", txID, err)
+					require.Contains(t, err.Error(), "Redirect blocked in test client: url:")
+					require.Contains(t, err.Error(), fmt.Sprintf("referrer: '%s/data/tx'", srv.URL()))
+				} else {
+					leaders[i] = 1
+					oldLeader = currentLeader
+					currentLeader = i
+					t.Logf("tx submitted: %s, %+v", txID, rcpt)
+					txOK++
+					t.Logf("Leader is: %s", srv.URL())
+					break
+				}
+			}
+		}
+		require.NoError(t, c.ShutdownServer(c.Servers[currentLeader]))
+		if oldLeader != -1 {
+			require.NoError(t, c.StartServer(c.Servers[oldLeader]))
+		}
+
+		time.Sleep(10 * time.Second)
```

Comment:
maybe it takes longer to elect a leader, so you can use the "eventually" approach, see above

Reply:
Done